common-ts/src/clients/marketDataFeed.ts

```diff
@@ -0,0 +1,710 @@
+
+	constructor(config: TradeSubscriptionConfig) {
+		super(config);
+		this.subject = new Subject<JsonTrade[]>();
+		this.subscriberSubscription = new TradeSubscriberSubscription(
+			this.id,
+			this.subject
+		);
+	}
+}
+
+type Subscriber = CandleSubscriber | TradeSubscriber;
+
+function getCompatibleCandleSubscriptionLookupKey(
+	config: Pick<CandleSubscriptionConfig, 'marketSymbol' | 'resolution' | 'env'>
+): CandleSubscriptionLookupKey {
+	return `${config.marketSymbol}:${config.resolution}:${config.env.key}` as CandleSubscriptionLookupKey;
+}
+
+function getCompatibleTradeSubscriptionLookupKey(
+	config: Pick<CandleSubscriptionConfig, 'marketSymbol' | 'env'>
+): TradeSubscriptionLookupKey {
+	return `${config.marketSymbol}:${config.env.key}` as TradeSubscriptionLookupKey;
+}
+
+class ApiSubscription {
+	public readonly id: ApiSubscriptionId;
+	public readonly candleSubscribers: Map<SubscriberId, CandleSubscriber>;
+	public readonly tradeSubscribers: Map<SubscriberId, TradeSubscriber>;
+	public readonly config: Pick<
+		CandleSubscriptionConfig,
+		'marketSymbol' | 'resolution' | 'env'
+	>;
+	private apiClient: DataApiWsClient;
+	private onNoMoreSubscribers: () => void;
+	private static idIncrementer: number = 0;
+
+	private generateSubscriptionId(): ApiSubscriptionId {
+		return `${getCompatibleCandleSubscriptionLookupKey(
+			this.config
+		)}(${ApiSubscription.idIncrementer++})` as ApiSubscriptionId;
+	}
+
+	get tradeSubscriptionsLookupKey() {
+		return getCompatibleTradeSubscriptionLookupKey(this.config);
+	}
+
+	get candleSubscriptionsLookupKey() {
+		return getCompatibleCandleSubscriptionLookupKey(this.config);
+	}
+
+	constructor(
+		marketSymbol: MarketSymbol,
+		resolution: CandleResolution,
+		env: UIEnv,
+		initialSubscriber: Subscriber,
+		onNoMoreSubscribers: (subscription: ApiSubscriptionId) => void
+	) {
+		this.config = { marketSymbol, resolution, env };
+		this.id = this.generateSubscriptionId();
+		this.onNoMoreSubscribers = () => onNoMoreSubscribers(this.id);
+		console.log(`marketDataFeed::creating_new_api_subscription:${this.id}`);
+
+		const initialSubscriberType = initialSubscriber.config.type;
+
+		this.apiClient = new DataApiWsClient({
+			marketSymbol: this.config.marketSymbol,
+			resolution: this.config.resolution,
+			env: this.config.env,
+		});
+
+		initialSubscriber.setSubscription(this);
+
+		switch (initialSubscriberType) {
+			case 'candles': {
+				this.candleSubscribers = new Map([
+					[initialSubscriber.id, initialSubscriber as CandleSubscriber],
+				]);
+				this.tradeSubscribers = new Map();
+				break;
+			}
+			case 'trades': {
+				this.candleSubscribers = new Map();
+				this.tradeSubscribers = new Map([
+					[initialSubscriber.id, initialSubscriber as TradeSubscriber],
+				]);
+				break;
+			}
+			default: {
+				const _never: never = initialSubscriberType;
+				throw new Error(`Unknown subscription type: ${_never}`);
+			}
+		}
+
+		this.subscribeToApi();
```

Comment:
i feel sus about this implementation...

we are subscribing to the API on object creation, but we don't have any information about the subscription from the consumer's POV.

e.g. we do `const marketDataApiSubscription = new ApiSubscription(...)`

but this code don't know anything about the underlying subscription, whereas we'd usually have something like `const subscription = await marketDataApiSubscription.subscribe()`

Reply:
I guess the idea was that apisubscription IS A SUBSCRIPTION .. it shouldn't exist without a subscriber/without being subscribed .. so it's got a minimal interface where the only public methods available are to add or remove subscribers (and it will automatically unsubscribe when there are no subscribers left) .. I'll wait to see what you think , I don't think it's that bad tbh 🧐